### ISSUE TITLE
Implement checkpointing to resume training

### ISSUE THREAD

**Issue by drewoldag** [lincc-frameworks/fibad#88]
We need to checkpoint during the training process.

- Capture both the latest checkpoint as well as the best checkpoint by some metric (loss, accuracy, etc)
- Want to be able to specify a checkpoint to restart from. 
- Want to be able to specify a checkpoint to continue training from. 

The checkpoint should include:
Optimizer state
Training loop state
Random number generator state
Model weights/state
